teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,538 @@
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+- What impact does this change have on the start-up time and execution time
+of TaskRuns and PipelineRuns?
+- What impact does it have on the resource footprint of Tekton controllers
+as well as TaskRuns and PipelineRuns?
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Tekton ecosystem. Consider including folks that also work outside the WGs
+or subproject.
+- How will security be reviewed and by whom?
+- How will UX be reviewed and by whom?
+-->
+
+### Drawbacks
+
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### Inputs / Outputs Distinguishment
+
+We can also separately collect inputs provenance from params, and outputs provenance from TaskRun results. The signable artifacts being found in params will be the inputs, and those found in TaskRun results will be the outputs. However, params are not reliable as users don't really specify those human less readable provenance metadata formats mentioned in in-toto attestations in neither PipelineRun nor TaskRun params.
+
+### Using Run Status to generate Provenance Metadata
+
+Results is not an ideal place to have the provenance metadata populated for the artifacts, since Results are provided by Pipeline authors. When people are looking for provenance, the metadata being provided should ideally be trustful and unchangable after generation. How can we garantee the provenance metadata can be trusted? Assuming the Tekton Pipeline being installed and where the Runs are operated are trustful, one missing piece to comply to SLSA level L3 is that the Run yaml being submitted is a trustful config, which can be fulfilled by [TEP 091: Trusted Resource](https://github.com/tektoncd/community/pull/739). The other missing piece is that we want to make sure the generated provenance data shouldn't be changable by any untrusted threads. TaskRun and PipelineRun `status` field satisfies these needs, as only Pipeline controller is able to generate info within this field. To extend the previous trustful setup, we can allow Trusted Tasks to modify the field as well: when Trusted Tasks are being used, Pipeline Controller can verify the Trusted Tasks being used in a Run, and let Trusted Tasks to generate provenance metadata a new field under `status` fields.
+
+This approach would require changes in Pipeline and also completion of [Trusted Resources](https://github.com/tektoncd/community/pull/739), so the detailed design can be scoped out in a future TEPs. However, the schemas for the provenance metadata should follow those defined in this TEP.
+
+## Implementation Plan
+
+The implementation plan has two main components:
+
+1. upgrading Tekton Chains's Tekton Pipeline's version to v0.38, which TEP 75 and 76 will be supported behind alpha feature flag.
```

Comment:
> Thank you for this!
> 
> Even with TEP-0075 and TEP-0076 implemented, it seems to me that we still lack the ability to express that a result is signable on Tekton Pipeline side.
> 
> I don't like very much the idea of asking Task authors to write the name of the result to a result in a script. The alternative would be to name results with a known prefix, like 'ARTIFACT_INPUT_something_something`. That wouldn't look very nice but it would save the extra result writing in the step.

Thanks for the comments! I think the TEP design was originally based on 'ARTIFACT_INPUT_something_something`, but this reduces the flexibility that Task authors has to follow the naming in a particular format. And that approach can already be applied through a merged implementation https://github.com/tektoncd/chains/pull/465/files. I think the TEP is trying to group multiple infos in a structured way, so I think having two extra results wouldn't hurt that much considering the flexibility in naming it provides.

Reply:
Thats right!